### PR TITLE
feat: add search on docs

### PR DESCRIPTION
**Pull request by tedim52**
## Description
- enables users to search through our docs
- with search, users should be able to navigate/find info in our docs much more easily

## User Facing
- [x] Yes

### CODE REVIEW

File: docs/docusaurus.config.js [hunk 1]
```diff
@@ -156,6 +156,27 @@ const config = {
         theme: lightCodeTheme,
         darkTheme: darkCodeTheme,
       },
+      algolia: {
+        appId: 'NTSX40VZB8',
+
+        // Public API key, safe to commit
```
Comment (h4ck3rk3y) [inline docs/docusaurus.config.js:168]: useful comment! :)

**Comment by leeederek** [comment]
Hey @tedim52, this is great. Search functionality on our docs is much needed! Two questions:
1. Is the theme of the search pop-up box something we can change? It looks like it might be specific to Algolia, but just wondering because the theme style doesn't exactly match the rest of the docs and website.
2. When testing functionality using [branch preview URL](https://tedi-docsearch.kurtosis-docs.pages.dev), I wasn't able to test out the workflow/experience of this search functionality. Do you know of a way I can test this before merging?
image

File: docs/docusaurus.config.js [hunk 2]
```diff
@@ -156,6 +156,27 @@ const config = {
...
+        apiKey: '4269c726c2fea4e6cddfeb9a21cd3d4e',
+
+        indexName: 'kurtosis',
+
+        contextualSearch: true,
+
+        externalUrlRegex: 'external\\.com|domain\\.com',
```
Comment (h4ck3rk3y) [inline docs/docusaurus.config.js]: what does this dO?

File: docs/docusaurus.config.js [hunk 3]
```diff
@@ -156,6 +156,27 @@ const config = {
...
+
+        replaceSearchResultPathname: {
```
Comment (h4ck3rk3y) [inline docs/docusaurus.config.js]: what does this do?

**Review by h4ck3rk3y — COMMENTED** [review]
Seems like search is tied into version so I got no result when I tried the page. @tedim52 lets see if we can index more often; have shared fears on Slack

Failure mode - Gyani releases new docs on Wednesday Morning
Angelo at Meta tries using Docs on Friday with Kurtosis but it isn't working, he can't find Starlark; he assumes no Starlark docs
On Tuesday things get indexed

File: docs/docusaurus.config.js (hunk 2 above)
Comment (tedim52) [reply docs/docusaurus.config.js]: `// Optional: Specify domains where the navigation should occur through window.location instead on history.push. Useful when our Algolia config crawls multiple documentation sites and we want to navigate with window.location.href to them.`

File: docs/docusaurus.config.js (hunk 3 above)
Comment (tedim52) [reply docs/docusaurus.config.js]: `// Optional: Replace parts of the item URLs from Algolia. Useful when using the same search index for multiple deployments using a different baseUrl. You can use regexp or string in the `from` param. For example: localhost:3000 vs myCompany.com/docs` - ended up removing cause I don't know if the options needed, can add back later if needed.

File: docs/docusaurus.config.js (hunk 2 above)
Comment (tedim52) [reply docs/docusaurus.config.js]: ended up removing cause I don't know if the options needed, can add back later if needed.

**Comment by tedim52** [comment]
@h4ck3rk3y - addressed your comment by adding a gh workflow that reindexes website everytime version docs are published.

@leeederek - addressed your comment by making the search box background, the same dark black color as website.

File: docs/src/css/custom.css [hunk 4]
```diff
@@ -52,4 +52,45 @@ main {
...
+  --docsearch-hit-active-color: var(--ifm-color-white);
+  --docsearch-hit-background: var(--ifm-color-emphasis-100);
+  /* Footer */
+  --docsearch-footer-background: var(--ifm-background-surface-color);
+  --docsearch-key-gradient: linear-gradient(
+          -26.5deg,
+          var(--ifm-color-emphasis-200) 0%,
+          var(--ifm-color-emphasis-100) 100%
+  );
 }
```
Comment (h4ck3rk3y) [inline docs/src/css/custom.css:96]: new line please

File: .github/workflows/push-docs.yml [hunk 5]
```diff
@@ -33,3 +33,17 @@ jobs:
...
+    runs-on: ubuntu-latest
+    name: Update website search index via Algolia Crawler
+    steps:
+      - uses: fjogeleit/http-request-action@v1
+        with:
+          url: 'https://crawler.algolia.com/api/1/crawlers/${{ secrets.ALGOLIA_CRAWLER_ID }}/reindex'
+          method: 'POST'
+          username: ${{ secrets.ALGOLIA_CRAWLER_USER_ID }}
+          password: ${{ secrets.ALGOLIA_CRAWLER_API_KEY }}
+          customHeaders: '{"Content-Type": "application/json"}'
```
Comment (h4ck3rk3y) [inline .github/workflows/push-docs.yml]: chatted on slack; we are going for reindex-main and reindex-yml

**Review by leeederek — APPROVED** [review]
Looks solid, thanks!

File: .github/workflows/reindex-website-main.yml [hunk 6]
```diff
@@ -0,0 +1,19 @@
+name: "Reindex website with Algolia Docsearch on main"
+on:
+  push:
+    branches:
+      - 'main'
+jobs:
+  reindex-website-main:
+    # Only reindex if a new version release is occurring, initiated by the release PR
+    if: contains(github.head_ref, 'release-please')
```
Comment (h4ck3rk3y) [inline .github/workflows/reindex-website-main.yml]: this doesn't exist on PRs

File: .github/workflows/reindex-website-main.yml [hunk 7]
```diff
@@ -0,0 +1,19 @@
+name: "Reindex website with Algolia Docsearch on main"
+on:
+  push:
+    branches:
+      - 'main'
+jobs:
+  reindex-website-main:
+    # Only reindex if a new version release is occurring, initiated by the release PR
```
Comment (h4ck3rk3y) [inline .github/workflows/reindex-website-main.yml]: off comment

File: .github/workflows/reindex-website-main.yml (hunk 6 above)
Comment (h4ck3rk3y) [reply .github/workflows/reindex-website-main.yml]: https://docs.github.com/en/actions/learn-github-actions/contexts

File: .github/workflows/reindex-website-main.yml (hunk 7 above)
Comment (h4ck3rk3y) [reply .github/workflows/reindex-website-main.yml]: the PR looks green; maybe its within comment and GH is being funny

File: .github/workflows/reindex-website-release.yml [hunk 8]
```diff
@@ -0,0 +1,19 @@
+name: "Reindex website with Algolia Docsearch for Releases"
+on:
+  push:
+    tags:
+      - '[0-9]+.[0-9]+.[0-9]+'
+jobs:
+  reindex-website-release:
+    # Only reindex if a new version release is occurring, initiated by the release PR
+    if: contains(github.head_ref, 'release-please')
```
Comment (h4ck3rk3y) [inline .github/workflows/reindex-website-release.yml]: don't think you need this at all

File: .github/workflows/reindex-website-main.yml (hunk 6 above)
Comment (h4ck3rk3y) [reply .github/workflows/reindex-website-main.yml]: dont think you need this at all

File: .github/workflows/reindex-website-main.yml (hunk 7 above)
Comment (tedim52) [reply .github/workflows/reindex-website-main.yml]: removed anyways haha

**Pull request merged**